File: example/1d-linear-convection/weno3/python-v1/02f/src/physics/problems/linear_advection.py
# src/physics/problems/linear_advection.py
"""
线性对流问题
"""

from .base import Problem
from physics.equations.linear_advection import LinearAdvectionEquation
from physics.systems.linear_advection_system import LinearAdvectionSystem
from physics.initial_conditions.factory import InitialConditionFactory
from core.registry import register_component

@register_component('problem', 'linear_advection')
class LinearAdvectionProblem(Problem):
    """
    线性对流问题：u_t + c u_x = 0
    使用周期边界 + 任意初始条件
    """

    def __init__(self, config):
        super().__init__(config)
        
    def initial_condition(self):
        """返回初始条件实例"""
        if self._initial_condition is None:
            from physics.initial_conditions.factory import InitialConditionFactory
            self._initial_condition = InitialConditionFactory.create(self.config)
        return self._initial_condition

    def physical_system(self):
        """返回物理系统实例"""
        if self._physical_system is None:
            # 创建方程
            equation = LinearAdvectionEquation(wave_speed=self.config.wave_speed)
            # 创建系统
            self._physical_system = LinearAdvectionSystem(equation)
        return self._physical_system

File: example/1d-linear-convection/weno3/julia/03g/src/equations/linear_advection.jl
# src/equations/linear_advection.jl

"""
线性对流方程: ∂u/∂t + c ∂u/∂x = 0
"""
mutable struct LinearAdvectionEquation <: Equation
    wave_speed::Float64
    function LinearAdvectionEquation(config::Any)
        c = getfield_safe(config, :wave_speed, 1.0)
        new(c)
    end
end

eq_flux(eq::LinearAdvectionEquation, u::Float64) = eq.wave_speed * u

max_wave_speed(eq::LinearAdvectionEquation, u::Float64) = abs(eq.wave_speed)